### PR TITLE
Validate API: provide more verbose explanation

### PR DESCRIPTION
**Pull request by alexksikes**
For Fuzzy Queries:

```
GET /imdb/movies/_validate/query?explain=true
{
  "query": {
    "fuzzy": {
      "actors": "kyle"
    }
  }
}

Response:

{
   ...
   "explanations": [
      {
         "index": "imdb",
         "valid": true,
         "explanation": "filtered(actors:eyle^0.75 actors:kale^0.75 actors:kayle^0.75 ... )->cache(_type:movies)"
      }
   ]
}
```

For More Like This:

```
GET /imdb/movies/_validate/query?explain=true
{
  "query": {
    "more_like_this": {
      "like": {
        "_id": "88247"
      }
    }
  }
}

Response:

{
   ...
   "explanations": [
      {
         "index": "imdb",
         "valid": true,
         "explanation": "filtered((((title:terminator^3.71334 plot:kyle^1.0604408 plot:cyborg^1.0863208 ... )~2)) -ConstantScore(_uid:movies#88247))->cache(_type:movies)"
      }
   ]
}
```

Relates to #1412

### CODE REVIEW

File: src/main/java/org/elasticsearch/action/admin/indices/validate/query/TransportValidateQueryAction.java [hunk 1]
```diff
@@ -188,7 +191,7 @@ protected ShardValidateQueryResponse shardOperation(ShardValidateQueryRequest re
 
             valid = true;
             if (request.explain()) {
-                explanation = searchContext.query().toString();
+                explanation = searchContext.query().rewrite(searcher.reader()).toString();
```
Comment (javanna) [inline src/main/java/org/elasticsearch/action/admin/indices/validate/query/TransportValidateQueryAction.java]: Not very expert on this aspect but I wonder if and how rewriting affects other queries explain output? Also, tests would help here just to validate what we print out.

**Comment by s1monw** [comment]
I think this LGTM - I wonder if we should make the rewrite optional and add a boolean to the request to trigger it?

**Comment by alexksikes** [comment]
I wondered this too if I should not add a verbose flag, but the slight problem is when the query is not valid, in this case, the explanation of the error is not more verbose. So the explain flag should explain the error if the query is not valid, and otherwise should explain the query itself if valid.

**Comment by s1monw** [comment]
just add `rewrite: true|false`?

**Comment by alexksikes** [comment]
`rewrite: true|false` would only make sense if `explain: true` then. I agree that this is more descriptive as to what is actually happening, but wondering if having to set two flags `explain: true` and `rewrite: true` is a good pattern for the task. Perhaps @clintongormley has some ideas.

**Comment by clintongormley** [comment]
I can imagine this output could be quite heavy, so I'd be happy having the extra `rewrite:true` flag.  The other place I'd love to see this is in the `common_terms` query and the `match` query when combined with `cutoff_frequency`.

**Comment by s1monw** [comment]
LGTM

**Pull request closed without merging**